### PR TITLE
ARROW-4560: [R] array() needs to take single input, not ...

### PR DESCRIPTION
**Pull request by romainfrancois**
This will simplify the handling of the `type` argument. 

jira: https://issues.apache.org/jira/browse/ARROW-4560?filter=12344983

### CODE REVIEW

**Comment by romainfrancois** [comment]
Also planning to work on https://issues.apache.org/jira/browse/ARROW-3810 in this PR

**Comment by romainfrancois** [comment]
To handle the `type=` argument in `array()` we need to be able to infer arrow types from R objects, so I've added the `type()` function: 

``` r
library(arrow, warn.conflicts = FALSE)

type(1:10)
#> arrow::Int32 
#> int32
type(1)
#> arrow::Float64 
#> double
type("")
#> arrow::Utf8 
#> string
type(iris$Species)
#> arrow::DictionaryType 
#> dictionary<values=string, indices=int8, ordered=0>
```

<sup>Created on 2019-02-14 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by romainfrancois** [comment]
I'm now fairly confident about this PR. `array()` used to take `...` and rely on `vctrs` to first combine to a common type using `vctrs` type system, though the `.ptype`. I've changed that so that `array()` only take one vector `x` and the `type=` argument is an arrow logical type, e.g. `int32()`. 

``` r
library(arrow, warn.conflicts = FALSE)

a <- array(1:10, type = int64())
a$type
#> arrow::Int64 
#> int64
a$as_vector()
#> integer64
#>  [1] 1  2  3  4  5  6  7  8  9  10

a <- array(1:10, type = float64())
a$type
#> arrow::Float64 
#> double
a$as_vector()
#>  [1]  1  2  3  4  5  6  7  8  9 10
```

The type of array that is made is governed by the `type=` argument. If missing, the type is inferred from the data, e.g. 

``` r
library(arrow, warn.conflicts = FALSE)

a <- array(rnorm(10))
a$type
#> arrow::Float64 
#> double
a$as_vector()
#>  [1]  0.48428301  0.74399694  0.48991163  0.35701469  0.60517996
#>  [6]  0.53463545  0.12579375  0.02414986 -0.49127583  0.17194012
```

The `chunked_array()` factory handles `...` and a type argument too: 

``` r
library(arrow, warn.conflicts = FALSE)

a <- chunked_array(rnorm(10), 1:10)
a$type
#> arrow::Float64 
#> double
a$as_vector()
#>  [1]  1.64642028 -1.47423016  0.84996187  1.21151724 -1.52303727
#>  [6] -0.04387242 -0.47798708 -0.18693768 -0.98903429  0.30376938
#> [11]  1.00000000  2.00000000  3.00000000  4.00000000  5.00000000
#> [16]  6.00000000  7.00000000  8.00000000  9.00000000 10.00000000

a <- chunked_array(1:5, 1:10, type = int64())
a$type
#> arrow::Int64 
#> int64
a$as_vector()
#> integer64
#>  [1] 1  2  3  4  5  1  2  3  4  5  6  7  8  9  10
```

File: r/tests/testthat/test-type.R [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+})
+
+test_that("type() gets the right type for arrow::ChunkedArray", {
+  a <- chunked_array(1:10, 1:10)
+  expect_equal(type(a), a$type)
+})
+
+test_that("type() infers from R type", {
+  expect_equal(type(1:10), int32())
+  expect_equal(type(1), float64())
```
Comment (wesm) [inline r/tests/testthat/test-type.R]: The optics of this are a bit odd =)

File: r/tests/testthat/test-type.R [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  expect_equal(type(TRUE), boolean())
+  expect_equal(type(raw()), int8())
```
Comment (wesm) [inline r/tests/testthat/test-type.R]: also slightly odd

File: r/tests/testthat/test-type.R [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  expect_equal(type(""), utf8())
+  expect_equal(
+    type(iris$Species),
+    dictionary(int8(), array(levels(iris$Species)), FALSE)
```
Comment (wesm) [inline r/tests/testthat/test-type.R]: maybe test for ordered factors at some point also?

**Review by wesm — APPROVED** [review]
+1, this looks nice. I'm going to rebase to fix any linting issue since I just added the cpplint checks

**Comment by wesm** [comment]
Oops I missed array_from_vector.cpp in my code review because it was hidden. silly GitHub. I'll take a skim through anyway

File: r/src/array_from_vector.cpp [hunk 4]
```diff
@@ -0,0 +1,1014 @@
...
+public:
+  virtual ~VectorConverter() = default;
+
+  virtual Status Init(ArrayBuilder* builder) = 0;
+
+  virtual Status Ingest(SEXP obj) = 0;
+
+  virtual Status GetResult(std::shared_ptr<arrow::Array>* result) {
+    RETURN_NOT_OK(builder_->Finish(result));
+    return Status::OK();
```
Comment (wesm) [inline r/src/array_from_vector.cpp]: you can `return builder_->Finish(result)`

**Review by wesm — COMMENTED** [review]
Some minor comments. I'll do a better job reviewing next time...

File: r/tests/testthat/test-type.R (hunk 1 above)
Comment (romainfrancois) [reply r/tests/testthat/test-type.R]: yeah maybe that can be `infer_type()` or something

File: r/src/array_from_vector.cpp [hunk 5]
```diff
@@ -0,0 +1,1014 @@
...
+  }
+
+  if (i < n) {
+    STOP_IF_NOT_OK(AllocateBuffer(BitUtil::BytesForBits(n), &null_buffer));
+    internal::FirstTimeBitmapWriter null_bitmap_writer(null_buffer->mutable_data(), 0, n);
+
+    // catch up
+    for (R_xlen_t j = 0; j < i; j++, null_bitmap_writer.Next()) {
+      null_bitmap_writer.Set();
+    }
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply r/src/array_from_vector.cpp]: Sure. We can revisit next time on this.

File: r/src/array_from_vector.cpp [hunk 6]
```diff
@@ -0,0 +1,1014 @@
...
+      } else {
+        null_bitmap_writer.Set();
+        *p_offset = current_offset += LENGTH(s);
+      }
+    }
+
+    null_bitmap_writer.Finish();
+  }
+
+  // ----- data buffer
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply r/src/array_from_vector.cpp]: This is so we only `AllocateBuffer` the buffer with the final size we need. The `LENGTH()` here is quick, R strings know their size: https://purrple.cat/blog/2018/03/05/strings-know-their-own-length/

Is the alternative to grow a buffer as we go ? Maybe there is a `StringArrayBuilder` I can use

File: r/src/array_from_vector.cpp (hunk 4 above)
Comment (romainfrancois) [reply r/src/array_from_vector.cpp]: Thanks. Will do in a next PR.

**Pull request closed without merging**